File: uploaddemo/src/business/galary/java/com/jw/galary/img/ui/CropActivity.java


package com.jw.galary.img.ui;

import android.app.ProgressDialog;
import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.net.Uri;
import android.os.Bundle;
import android.os.Environment;
import android.support.annotation.Nullable;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.Button;
import android.widget.TextView;

import com.jw.galary.img.ImagePicker;
import com.jw.galary.img.crop.CropIwaView;
import com.jw.galary.img.crop.CropIwaView.CropSaveCompleteListener;
import com.jw.galary.img.crop.CropIwaView.ErrorListener;
import com.jw.galary.img.crop.config.CropIwaSaveConfig.Builder;
import com.jw.galary.img.crop.shape.CropIwaOvalShape;
import com.jw.galary.img.view.CropImageView.Style;
import com.jw.uploaddemo.ColorCofig;
import com.jw.uploaddemo.R;
import com.jw.uploaddemo.uploadPlugin.UploadPluginActivity;

import java.io.File;

public class CropActivity extends UploadPluginActivity implements OnClickListener {
    private static final String EXTRA_URI = "CropImage";
    private ImagePicker imagePicker;
    private CropIwaView cropView;
    private ProgressDialog mProgressDialog;
    private String dstPath;
    private float originAngle;
    private View topBar;
    private View bottomBar;

    public CropActivity() {
    }

    public static Intent callingIntent(Context context, Uri imageUri) {
        Intent intent = new Intent(context, CropActivity.class);
        intent.putExtra("CropImage", imageUri);
        return intent;
    }

    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        this.setContentView(R.layout.activity_crop);
        this.findViewById(R.id.tv_rotate).setOnClickListener(this);
        this.findViewById(R.id.tv_recover).setOnClickListener(this);
        this.findViewById(R.id.btn_back).setOnClickListener(this);
        Button mBtnOk = this.findViewById(R.id.btn_ok);
        mBtnOk.setText(this.getString(R.string.ip_complete));
        mBtnOk.setOnClickListener(this);
        this.imagePicker = ImagePicker.INSTANCE;
        Uri imageUri = this.getIntent().getParcelableExtra("CropImage");
        this.cropView = this.findViewById(R.id.cv_crop_image);
        this.cropView.setImageUri(imageUri);
        this.cropView.configureOverlay().setAspectRatio(this.imagePicker.getAspectRatio()).setDynamicCrop(this.imagePicker.isDynamicCrop()).apply();
        this.cropView.setOnClickListener(this);
        if (this.imagePicker.getStyle() == Style.CIRCLE) {
            this.cropView.configureOverlay().setCropShape(new CropIwaOvalShape(this.cropView.configureOverlay())).apply();
        }

        File cropCacheFolder;
        if (this.imagePicker.getCutType() == 2) {
            cropCacheFolder = new File(Environment.getExternalStorageDirectory() + "/RXImagePicker/");
        } else {
            cropCacheFolder = ImagePicker.INSTANCE.getCropCacheFolder();
        }

        if (!cropCacheFolder.exists() || !cropCacheFolder.isDirectory()) {
            cropCacheFolder.mkdirs();
        }

        this.dstPath = (new File(cropCacheFolder, "IMG_" + System.currentTimeMillis() + ".png")).getAbsolutePath();
        this.cropView.setCropSaveCompleteListener(new CropSaveCompleteListener() {
            public void onCroppedRegionSaved(Uri bitmapUri) {
                CropActivity.this.dismiss();
                ImagePicker.INSTANCE.galleryAddPic(CropActivity.this.getApplicationContext(), bitmapUri);
                Intent intent = new Intent();
                intent.putExtra(ImagePicker.EXTRA_CROP_IMAGE_OUT_URI, bitmapUri);
                CropActivity.this.setResult(-1, intent);
                CropActivity.this.finish();
            }
        });
        this.cropView.setErrorListener(new ErrorListener() {
            public void onError(Throwable e) {
                CropActivity.this.dismiss();
            }
        });
        this.originAngle = this.cropView.getMatrixAngle();
        this.setConfirmButtonBg(mBtnOk);
        topBar = this.findViewById(R.id.top_bar);
        bottomBar = this.findViewById(R.id.bottom_bar);
        topBar.setBackgroundColor(Color.parseColor(ColorCofig.INSTANCE.getNaviBgColor()));
        ((TextView) this.findViewById(R.id.tv_des)).setTextColor(Color.parseColor(ColorCofig.INSTANCE.getNaviTitleColor()));
    }

    private void dismiss() {
        if (this.mProgressDialog != null && this.mProgressDialog.isShowing()) {
            this.mProgressDialog.dismiss();
        }

    }

    public void onClick(View v) {
        int id = v.getId();
        if (id == R.id.btn_back) {
            this.setResult(0);
            this.finish();
        } else if (id == R.id.tv_rotate) {
            this.cropView.rotateImage(90.0F);
        } else if (id == R.id.tv_recover) {
            float currentAngle = this.cropView.getMatrixAngle();
            if (this.originAngle != currentAngle) {
                this.cropView.rotateImage(this.originAngle - currentAngle);
            }

            this.cropView.initialize();
        } else if (id == R.id.btn_ok) {
            if (this.mProgressDialog == null) {
                this.mProgressDialog = new ProgressDialog(this);
                this.mProgressDialog.setMessage("正在处理中...");
                this.mProgressDialog.setCanceledOnTouchOutside(false);
                this.mProgressDialog.setCancelable(false);
            }

            this.mProgressDialog.show();
            Builder builder = new Builder(Uri.fromFile(new File(this.dstPath)));
            if (this.imagePicker.getOutPutX() != 0 && this.imagePicker.getOutPutY() != 0 && !this.imagePicker.isOrigin()) {
                builder.setSize(this.imagePicker.getOutPutX(), this.imagePicker.getOutPutY());
            }

            builder.setQuality(this.imagePicker.getQuality());
            this.cropView.crop(builder.build());
        }
    }

    protected void onDestroy() {
        super.onDestroy();
        this.dismiss();
        this.mProgressDialog = null;
    }
}

File: uploaddemo/src/business/galary/java/com/jw/galary/img/ui/CropActivity.kt
package com.jw.galary.img.ui

import android.app.ProgressDialog
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.net.Uri
import android.os.Bundle
import android.os.Environment
import android.view.View
import android.view.View.OnClickListener
import android.widget.TextView
import com.jw.galary.img.ImagePicker
import com.jw.galary.img.crop.CropIwaView
import com.jw.galary.img.crop.config.CropIwaSaveConfig.Builder
import com.jw.galary.img.crop.shape.CropIwaOvalShape
import com.jw.galary.img.view.CropImageView.Style
import com.jw.uploaddemo.ColorCofig
import com.jw.uploaddemo.R
import com.jw.uploaddemo.databinding.ActivityCropBinding
import com.jw.uploaddemo.uploadPlugin.UploadPluginBindingActivity
import kotlinx.android.synthetic.main.activity_crop.*
import kotlinx.android.synthetic.main.include_top_bar.view.*
import java.io.File

class CropActivity : UploadPluginBindingActivity<ActivityCropBinding>(),
    CropIwaView.CropSaveCompleteListener, CropIwaView.ErrorListener {
    private var mProgressDialog: ProgressDialog? = null
    private var dstPath: String? = null
    private var originAngle: Float = 0.toFloat()

    override fun getLayoutId() = R.layout.activity_crop

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mBinding.apply {
            clickListener = OnClickListener {
                when (it.id) {
                    R.id.tv_rotate -> rotate()
                    R.id.tv_recover -> recover()
                    R.id.btn_back -> back()
                    R.id.btn_ok -> crop()
                }
            }
            top_bar.btn_ok.text = getString(R.string.ip_complete)
        }

        val imageUri = intent.getParcelableExtra<Uri>("CropImage")
        mBinding.cvCropImage.setImageUri(imageUri)
        mBinding.cvCropImage.configureOverlay().setAspectRatio(ImagePicker.aspectRatio)
            .setDynamicCrop(ImagePicker.isDynamicCrop).apply()
        if (ImagePicker.style == Style.CIRCLE) {
            mBinding.cvCropImage.configureOverlay()
                .setCropShape(CropIwaOvalShape(mBinding.cvCropImage.configureOverlay())).apply()
        }

        val cropCacheFolder = if (ImagePicker.cutType == 2) {
            File(Environment.getExternalStorageDirectory().toString() + "/RXImagePicker/")
        } else {
            ImagePicker.cropCacheFolder
        }

        if (!cropCacheFolder!!.exists() || !cropCacheFolder.isDirectory) {
            cropCacheFolder.mkdirs()
        }

        dstPath =
            File(cropCacheFolder, "IMG_" + System.currentTimeMillis() + ".png").absolutePath
        mBinding.cvCropImage.setCropSaveCompleteListener(this)
        mBinding.cvCropImage.setErrorListener { this@CropActivity.dismiss() }
        originAngle = mBinding.cvCropImage.matrixAngle
        setConfirmButtonBg(top_bar.btn_ok)
        top_bar.setBackgroundColor(Color.parseColor(ColorCofig.naviBgColor))
        (findViewById<View>(R.id.tv_des) as TextView).setTextColor(Color.parseColor(ColorCofig.naviTitleColor))
    }

    private fun rotate() {
        mBinding.cvCropImage.rotateImage(90.0f)
    }

    private fun recover() {
        val currentAngle = mBinding.cvCropImage.matrixAngle
        if (originAngle != currentAngle) {
            mBinding.cvCropImage.rotateImage(originAngle - currentAngle)
        }

        mBinding.cvCropImage.initialize()
    }

    private fun crop() {
        if (mProgressDialog == null) {
            mProgressDialog = ProgressDialog(this@CropActivity)
            mProgressDialog!!.setMessage("正在处理中...")
            mProgressDialog!!.setCanceledOnTouchOutside(false)
            mProgressDialog!!.setCancelable(false)
        }

        mProgressDialog!!.show()
        val builder = Builder(Uri.fromFile(File(dstPath)))
        if (ImagePicker.outPutX != 0 && ImagePicker.outPutY != 0 && !ImagePicker.isOrigin) {
            builder.setSize(ImagePicker.outPutX, ImagePicker.outPutY)
        }

        builder.setQuality(ImagePicker.quality)
        mBinding.cvCropImage.crop(builder.build())
    }

    private fun back() {
        setResult(0)
        finish()
    }

    override fun onCroppedRegionSaved(bitmapUri: Uri?) {
        this@CropActivity.dismiss()
        ImagePicker.galleryAddPic(this@CropActivity.applicationContext, bitmapUri!!)
        val intent = Intent()
        intent.putExtra(ImagePicker.EXTRA_CROP_ITEM_OUT_URI, bitmapUri)
        this@CropActivity.setResult(ImagePicker.RESULT_CODE_ITEM_CROP, intent)
        this@CropActivity.finish()
    }

    override fun onError(var1: Throwable?) {
        dismiss()
    }

    private fun dismiss() {
        if (mProgressDialog != null && mProgressDialog!!.isShowing) {
            mProgressDialog!!.dismiss()
        }

    }

    override fun onDestroy() {
        super.onDestroy()
        dismiss()
        mProgressDialog = null
    }

    companion object {
        private val EXTRA_URI = "CropImage"

        fun callingIntent(context: Context, imageUri: Uri): Intent {
            val intent = Intent(context, CropActivity::class.java)
            intent.putExtra("CropImage", imageUri)
            return intent
        }
    }
}
